results/pkg/api/server/server.go

```diff
@@ -155,54 +163,186 @@ func (s *Server) ListResultsResult(ctx context.Context, req *pb.ListResultsReque
+	base64.RawURLEncoding.Encode(encodedResult, result)
+	return base64.RawURLEncoding.EncodeToString(encodedResult)
+}
 
-	// filter from all results
-	prg, err := s.env.Program(ast)
+func (s Server) decodeTokenToPrimaryKey(token string) (uint64, error) {
+	encodedToken, err := base64.RawURLEncoding.DecodeString(token)
+	if err != nil {
+		return 0, err
+	}
+	result := make([]byte, base64.RawURLEncoding.DecodedLen(len(encodedToken)))
+	_, err = base64.RawURLEncoding.Decode(result, encodedToken)
 	if err != nil {
-		log.Printf("program construction error: %s", err)
-		return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter checking step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
-	}
-	var resp []*pb.Result
-	for _, r := range results {
-		for _, e := range r.Executions {
-			out, _, err := prg.Eval(map[string]interface{}{
-				"taskrun": e.GetTaskRun(),
-			})
-			if err != nil {
-				log.Printf("failed to evaluate the expression: %v", err)
-				return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		return 0, err
+	}
+	p := &pb.Result{}
+	if err := proto.Unmarshal(result, p); err != nil {
+		log.Printf("unmarshaling error: %v", err)
+		return 0, status.Errorf(codes.Internal, "failed to parse result data: %v", err)
+	}
+	var primaryKey uint64
+	if err := s.db.QueryRow("SELECT id FROM results where name=?", p.GetName()).Scan(&primaryKey); err != nil {
+		return 0, err
+	}
+	return primaryKey, nil
+}
+
+// This function aims to obtain a fixed size `pageSize` of results from the database, starting
+// from the results with the primary key `startPK`, filtered by a compiled CEL program `prg`.
+//
+// In this function, we query the database multiple times and filter the queried results to
+// comprise the final results.
+//
+// To minimize the query times, we introduce a variable `ratio` to indicate the retention rate
+// after filtering a batch of results. The ratio of the queried batch is:
+//             ratio = remained_results_size/batch_size.
+//
+// The batchSize depends on the `ratio` of the previous batch and the `pageSize`:
+//                  batchSize = pageSize/last_ratio
+// The less the previous ratio is, the bigger the upcoming batch_size is. Then the queried time
+// is significantly decreased.
+func (s Server) getFilteredPaginatedResults(pageSize int32, startPK uint64, prg cel.Program) (results []*pb.Result, lastPK uint64, err error) {
+	if startPK == 0 {
+		lastPK = startPK
+	} else {
+		lastPK = startPK - 1
+	}
+	var ratio float32 = 1 // the retention rate after filtering a batch of results.
```

Comment:
it's not immediately obvious to me what retention rate means - we should describe it in more detail.